### DISCUSSION TITLE
replace logic and optimistic locking in general

### DISCUSSION THREAD

**Discussion by shawkins** [fabric8io/kubernetes-client#4137] [General]
We had a discussion about how optimistic locking is performed in general and the expectations for client methods.

Some possible topics that is spawned are:

- kubectl replace --force was the rationale for the deleteExisting support we had, but that has been removed - do we want to add some support back now that we can support a better blocking delete routine?

- digging through the rest of the kubectl replace implementation:
  - it does not do logic like modifyItemForReplaceOrPatch, for example the ClusterIP is not updated.  Moving forward things like that should not be needed https://github.com/kubernetes/kubernetes/pull/103532 
  - there's no locking flag.  If the resource has a resourceVerion, then that is used.  If none is supplied, then it will fetch latest.
  - it does not perform a retry, just a single attempt at getting the latest resourceVersion

Should we simplify/modify the replace routine?
- Deprecate modifyItemForReplaceOrPatch so that its usage does not proliferate.  If a user hits a similar problem on a legacy version it's better to push them towards a merge, apply, or strategic merge patch.  
- Remove retry
- Ideally match the kubectl resourceVersion behavior, but that is a breaking change.  It would make lockResourceVerion().replace() necessary.  It also make things more consistent with patch behavior.
 
On patch behavior: 
- a patchStatus() call was added.  But there is no patch(), nor patch(context) calls - those should probably be added. #4142
- the javadoc for patch(item) is currently wrong wrt to the implementation - it does not matter if you have used a load or a resource call, the patch base is always the latest state from the server. #4142
- kubectl edit will use either strategic merge or merge, not json.  We should have support for generating merge and strategic merge patches - however the implementation support for strategic merge is quite involved https://github.com/kubernetes/apimachinery/blob/master/pkg/util/strategicpatch/patch.go

**Comment by manusa** [comment]
_WIP_

Relates to: https://github.com/fabric8io/kubernetes-client/issues/4142
